### PR TITLE
#54 Non-interactive mode for MasterController

### PR DESCRIPTION
**Pull request by h-arlt**

### CODE REVIEW

File: src/main/java/com/xceptance/xlt/mastercontroller/Main.java [hunk 1]
```diff
@@ -422,21 +443,75 @@ private static CommandLine parseCommandLine(final String[] args)
...
+        System.out.println("    mastercontroller [<other options>]");
+        System.out.println("      -> Runs in interactive mode. Choose the command to be executed next.");
         System.out.println();
-        formatter.printHelp(XltConstants.MASTERCONTROLLER_EXECUTABLE_NAME + " [<options>]", "\nOptions:", options, "");
+        System.out.println("    mastercontroller -c <commandList> [<other options>]");
+        System.out.println("      -> Runs in non-interactive mode. Pass the commands to be executed.");
+        System.out.println();
+        System.out.println("    mastercontroller -auto [<other options>]");
+        System.out.println("      -> Runs a load test in non-interactive mode by executing all needed commands automatically.");
+        System.out.println();
```
Comment (h-arlt) [inline src/main/java/com/xceptance/xlt/mastercontroller/Main.java:456]: Why not using the formatter for printing that as was done before? It's a little bit weird to format and print the program usage text manually but use the formatter to format and print the program's options.

File: src/main/java/com/xceptance/xlt/mastercontroller/Main.java [hunk 2]
```diff
@@ -422,21 +443,75 @@ private static CommandLine parseCommandLine(final String[] args)
...
+     */
+    private static void validateCommandLine(final CommandLine commandLine)
+    {
+        boolean invalid = false;
+
+        // -auto|-faf|-sequential and -c are mutually exclusive
+        if ((commandLine.hasOption(OPTION_AUTO) || commandLine.hasOption(OPTION_FAF) || commandLine.hasOption(OPTION_SEQUENTIAL)) &&
+            commandLine.hasOption(OPTION_COMMANDS))
+        {
+            System.out.printf("Option '-%s' cannot be used together with '-%s', '-%s', or '-%s'.\n", OPTION_COMMANDS, OPTION_AUTO,
```
Comment (h-arlt) [inline src/main/java/com/xceptance/xlt/mastercontroller/Main.java:483]: Log at level ERROR additionally?

File: src/main/java/com/xceptance/xlt/mastercontroller/Main.java [hunk 3]
```diff
@@ -422,21 +443,75 @@ private static CommandLine parseCommandLine(final String[] args)
...
+
+        // check -c option value
+        final String commandList = commandLine.getOptionValue(OPTION_COMMANDS);
+        if (commandList != null)
+        {
+            final String[] unknownCommands = MasterControllerCommands.validate(commandList);
+            if (unknownCommands.length > 0)
+            {
+                System.out.printf("Unknown commands given with the '-%s' option: %s\n", OPTION_COMMANDS,
+                                  StringUtils.join(unknownCommands, ", "));
```
Comment (h-arlt) [inline src/main/java/com/xceptance/xlt/mastercontroller/Main.java:496]: Log at level ERROR additionally?

**Review by h-arlt — COMMENTED** [review]
Besides a few small improvements/suggestions overall code looks good to me.

Q: Do we have to point out (in the help text, for example) that the passed commands are executed in the same order as given on the command line (even if they don't make sense)?

File: src/main/java/com/xceptance/xlt/mastercontroller/Main.java (hunk 1 above)
Comment (jowerner) [reply src/main/java/com/xceptance/xlt/mastercontroller/Main.java:456]: True. Fixed.

**Comment by jowerner** [comment]
> Q: Do we have to point out (in the help text, for example) that the passed commands are executed in the same order as given on the command line (even if they don't make sense)?

Yep. Added this to the help text.

File: src/main/java/com/xceptance/xlt/mastercontroller/Main.java (hunk 2 above)
Comment (jowerner) [reply src/main/java/com/xceptance/xlt/mastercontroller/Main.java]: Done.

**Comment by h-arlt** [comment]
Merge at will ;-)

**Pull request merged**